File: xcount/events/stock_reconciliation.py
from __future__ import unicode_literals
import frappe
import json
from erpnext.stock.doctype.batch.batch import get_batch_qty, get_batches


def validate_items_and_stock_sheets(doc, method):
    #Validate if stock_sheets are not reconciled yet
    for row in doc.stock_sheets:
        print(row.stock_sheet)
        stock_sheet = frappe.get_doc('Stock Sheet', row.stock_sheet)
        if stock_sheet.reconciled == 1:
            frappe.throw('Stock Sheet# {0} is already reconciled'.format(stock_sheet.name))
        else:
            stock_sheet.db_set('reconciled', 1)

@frappe.whitelist()
def get_items_from_stocksheets(stock_sheets, warehouse, make_uncounted_items_zero = None):
    #Merge quantity of same items from same warehouse in stock_sheets
    stock_sheets  = json.loads(stock_sheets)
    items = {}
    items_set = set()
    for row in stock_sheets:
        stock = frappe.get_doc('Stock Sheet', row)
        for item in stock.items:
            items_set.add(item.item_code)
            item_tuple = (item.item_code, item.warehouse, item.batch_no)
            if not item_tuple in items:
                items[item_tuple] = item.qty
            else:
                items[item_tuple] += item.qty
    item_array = []
    for row in items:
        item_array.append({
            'item_code': row[0],
            'warehouse': row[1],
            'qty': items[row],
            'batch_no': row[2]
        })

    if not item_array: 
        return item_array
    
    if make_uncounted_items_zero:
        bin_list = frappe.get_list('Bin', filters = {'warehouse': warehouse}, fields = ['item_code'])
        for row in bin_list:
            if row.item_code not in items_set:

                has_batch = frappe.get_value('Item', row.item_code, 'has_batch_no')
                if has_batch: 
                    batches = get_batches(row.item_code, warehouse)
                    for batch in batches:
                        if batch.qty > 0:
                            item_array.append({
                                'item_code': row.item_code,
                                'warehouse': warehouse,
                                'qty': 0,
                                'batch_no': batch.batch_id
                            })
                else:
                    item_array.append({
                    'item_code': row.item_code,
                    'warehouse': warehouse,
                    'qty': 0
                })
    return item_array

File: xcount/public/js/stock_reconciliation.js
frappe.ui.form.on('Stock Reconciliation', {
    refresh: function (frm) {
        frm.events.purpose(frm);
    },
    purpose: function (frm) {  
        console.log(frm.doc.purpose);
        if (frm.doc.purpose === 'Stock Reconciliation'){
            frm.events.create_stock_sheet_button(frm);
        }
    },
    create_stock_sheet_button: function (frm) {  
        if (frm.doc.docstatus === 0){
            frm.add_custom_button(__("Get Stock Count Sheets"), function() {
                if (!frm.doc.warehouse){
                    frappe.msgprint('No Warehouse selected')
                }else{
                    frm.trigger('get_stock_sheets');
                }
			});
        }
    },

    get_stock_sheets: function (frm) {
        frappe.db.get_list('Stock Sheet', {
            fields: ['name', 'stock_count_date', 'default_warehouse'],
            filters: {
                reconciled: 0,
                docstatus: 1
            }
        }).then(records => {
            if (records.length === 0){
                frappe.msgprint(`No Stock Count Sheet pending for ${frm.doc.warehouse}`);
            }
            frm.doc.stock_sheets = []
            frm.doc.items = []
            for (let row in records) {
                frm.add_child('stock_sheets', {
                    'stock_sheet': records[row].name,
                    'stock_count_date': records[row].stock_count_date
                })
            }

            frm.refresh_field('stock_sheets');

            frappe.call('xcount.events.stock_reconciliation.get_items_from_stocksheets', 
            {stock_sheets : records})
            .then(r => {
                for (let row in r.message) {
                    frm.add_child('items', {
                        'item_code': r.message[row].item_code,
                        'warehouse': r.message[row].warehouse,
                        'qty': r.message[row].qty,
                        'batch_no': r.message[row].batch_no
                    })
                }
                frm.refresh_field('items');
            })
        })
        
    },
    set_batch_query: function (frm, cdt, cdn) {  
        let row = frappe.get_doc(cdt, cdn)
        if (row.item_code && row.warehouse){
            row.batch_no = null;
            frm.refresh_field('items')
            frm.set_query('batch_no', 'items', () => {
                return {
                    filters: {
                        item  : row.item_code
                    }
                }
            })
        }
    }
})

frappe.ui.form.on('Stock Reconciliation Item', {
    item_code: function (frm, cdt, cdn) {  
        frm.events.set_batch_query(frm, cdt, cdn);

    },
    warehouse: function (frm, cdt, cdn) {  
        frm.events.set_batch_query(frm, cdt, cdn);
    }
})